diff -up kdebase-workspace-4.3.4/plasma/dataengines/calendar/CMakeLists.txt.calendar-akonadi kdebase-workspace-4.3.4/plasma/dataengines/calendar/CMakeLists.txt
--- kdebase-workspace-4.3.4/plasma/dataengines/calendar/CMakeLists.txt.calendar-akonadi	2009-04-28 15:46:20.000000000 +0200
+++ kdebase-workspace-4.3.4/plasma/dataengines/calendar/CMakeLists.txt	2010-01-05 12:52:18.724518279 +0100
@@ -16,10 +16,9 @@ kde4_add_plugin(plasma_engine_calendar $
 target_link_libraries(
     plasma_engine_calendar
     kholidays
-    ${KDE4_KDECORE_LIBS} ${KDE4_AKONADI_LIBS}
+    ${KDE4_KDECORE_LIBS}
     ${KDE4_PLASMA_LIBS}
     ${KDE4_KMIME_LIBS}
-    ${KDE4_AKONADI_KMIME_LIBS}
 )
 
 install(TARGETS plasma_engine_calendar DESTINATION ${PLUGIN_INSTALL_DIR})
